File: worker/src/company_methods/specs.py
"""Versioned method-spec loader with repository-local path resolution."""

from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

from .schemas import CompanyMethodsError

SPEC_ROOT = Path(__file__).resolve().parents[3] / "method-specs"
ALLOWED_SPECS = frozenset(
    {"manifest", "hazards", "risk-score", "measurable-factors", "owas", "ejms", "chemical-inhalation", "source-anomalies"}
)


@lru_cache(maxsize=None)
def load_spec(name: str) -> dict[str, Any]:
    if name not in ALLOWED_SPECS:
        raise CompanyMethodsError(f"Unknown method spec: {name}")
    path = SPEC_ROOT / f"{name}.json"
    try:
        document = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, UnicodeError, json.JSONDecodeError) as error:
        raise CompanyMethodsError(f"Cannot read method spec {path.name}") from error
    if not isinstance(document, dict) or not isinstance(document.get("version"), str):
        raise CompanyMethodsError(f"Invalid method spec {path.name}")
    return document

File: src/lib/company-methods/specs.ts
import chemicalSpec from "../../../method-specs/chemical-inhalation.json" with { type: "json" };
import hazardSpec from "../../../method-specs/hazards.json" with { type: "json" };
import manifestSpec from "../../../method-specs/manifest.json" with { type: "json" };
import measurableSpec from "../../../method-specs/measurable-factors.json" with { type: "json" };
import owasSpec from "../../../method-specs/owas.json" with { type: "json" };
import riskScoreSpec from "../../../method-specs/risk-score.json" with { type: "json" };

export const companyMethodSpecs = {
  manifest: manifestSpec,
  hazards: hazardSpec,
  riskScore: riskScoreSpec,
  measurable: measurableSpec,
  owas: owasSpec,
  chemical: chemicalSpec,
} as const;

export const hazardSuggestions = hazardSpec.hazards.map((item) => ({
  id: item.id,
  label: item.display_label,
  source: item.source,
  effects: item.effects,
}));
